### PR TITLE
[ffmpeg] Modernize

### PR DESCRIPTION
**Pull request by MartinDelille**
- [ffmpeg/5.1] Add version
- Don't patch libx264.c from 5.1

Specify library name and version:  **lib/1.0**

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by MartinDelille** [comment]
@SpaceIm I have this PR pending since 2 weeks and wondered if you would accept to review it ?

**Comment by MartinDelille** [comment]
Hi @sorny92 ! Would you review my PR ? If I understand correctly, it might drag maintainer attention if I already have one community review.

**Comment by SpaceIm** [comment]
@MartinDelille you have to fix this import first, linter complains that you are importing a private function.

File: recipes/ffmpeg/all/conanfile.py [hunk 1]
```diff
@@ -313,23 +311,23 @@ def validate(self):
...
 
     def build_requirements(self):
         if self.settings.arch in ("x86", "x86_64"):
-            self.build_requires("yasm/1.3.0")
-        self.build_requires("pkgconf/1.9.3")
-        if self._settings_build.os == "Windows" and not tools.get_env("CONAN_BASH_PATH"):
-            self.build_requires("msys2/cci.latest")
+            self.tool_requires("yasm/1.3.0")
+        self.tool_requires("pkgconf/1.9.3")
+        if self._settings_build.os == "Windows" and not self.win_bash:
```
Comment (SpaceIm) [inline recipes/ffmpeg/all/conanfile.py:321]:
```suggestion
        if self._settings_build.os == "Windows" and not tools.get_env("CONAN_BASH_PATH"):
```

recipe is not migrated to conan v2 build helpers, so you have to keep this logic. Moreover testing self.win_bash in conan v2 is not correct, it's not set externally .

File: recipes/ffmpeg/all/conanfile.py [hunk 2]
```diff
@@ -382,15 +379,16 @@ def _configure_autotools(self):
         if self._autotools:
             return self._autotools
         self._autotools = AutoToolsBuildEnvironment(
-            self, win_bash=tools.os_info.is_windows)
+            self, win_bash=(self.settings.os == "Windows"))
```
Comment (SpaceIm) [inline recipes/ffmpeg/all/conanfile.py:382]:
```suggestion
            self, win_bash=(self._settings_build.os == "Windows"))
```

File: recipes/ffmpeg/all/test_package/conanfile.py [hunk 3]
```diff
@@ -1,29 +1,29 @@
...
     def build(self):
         cmake = CMake(self)
         cmake.configure()
         cmake.build()
 
     def test(self):
-        if not tools.cross_building(self):
+        if can_run(self):
             if self.options["ffmpeg"].with_programs:
                 self.run("ffmpeg --help", run_environment=True)
```
Comment (SpaceIm) [inline recipes/ffmpeg/all/test_package/conanfile.py:26]:
```suggestion
                self.run("ffmpeg --help", env="conanrun")
```

File: recipes/ffmpeg/all/test_v1_package/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    target_link_libraries(${PROJECT_NAME} ffmpeg::swscale)
+endif ()
+if (TARGET ffmpeg::swresample)
+    target_compile_definitions(${PROJECT_NAME} PRIVATE HAVE_FFMPEG_SWRESAMPLE)
+    target_link_libraries(${PROJECT_NAME} ffmpeg::swresample)
+endif ()
+if (TARGET ffmpeg::postproc)
+    target_compile_definitions(${PROJECT_NAME} PRIVATE HAVE_FFMPEG_POSTPROC)
+    target_link_libraries(${PROJECT_NAME} ffmpeg::postproc)
+endif ()
```
Comment (SpaceIm) [inline recipes/ffmpeg/all/test_v1_package/CMakeLists.txt:38]:
```suggestion
cmake_minimum_required(VERSION 3.1)
project(test_package)

include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
conan_basic_setup(TARGETS)

add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/../test_package
                 ${CMAKE_CURRENT_BINARY_DIR}/test_package)
```

File: recipes/ffmpeg/all/conanfile.py [hunk 5]
```diff
@@ -339,41 +337,40 @@ def _target_arch(self):
...
                                   "check_lib openssl openssl/ssl.h SSL_library_init -lssl -lcrypto -lws2_32 -lgdi32 ||",
-                                  "check_lib openssl openssl/ssl.h OPENSSL_init_ssl %s || " % openssl_libraries)
+                                  f"check_lib openssl openssl/ssl.h OPENSSL_init_ssl {openssl_libraries} || ")
 
     @contextlib.contextmanager
     def _build_context(self):
-        with tools.environment_append({"PKG_CONFIG_PATH": tools.unix_path(self.build_folder)}):
-            if self._is_msvc:
-                with tools.vcvars(self):
+        with environment_append({"PKG_CONFIG_PATH": unix_path(self, self.build_folder)}):
```
Comment (SpaceIm) [inline recipes/ffmpeg/all/conanfile.py:371]:
```suggestion
        with environment_append({"PKG_CONFIG_PATH": unix_path(self, self.generators_folder)}):
```

Comment (SpaceIm) [reply recipes/ffmpeg/all/conanfile.py:371]: I don't think it's a good idea to use unix_path when recipe is not migrated to conan v2 build helpers.

File: recipes/ffmpeg/all/conanfile.py [hunk 6]
```diff
@@ -537,55 +535,61 @@ def opt_append_disable_if_set(args, what, v):
...
-            if self.settings.compiler == "Visual Studio" and tools.Version(self.settings.compiler.version) <= "12":
+            if self.settings.compiler == "Visual Studio" and Version(self.settings.compiler.version) <= "12":
                 # Visual Studio 2013 (and earlier) doesn't support "inline" keyword for C (only for C++)
                 self._autotools.defines.append("inline=__inline")
         if cross_building(self):
             if self._target_os == "emscripten":
                 args.append("--target-os=none")
             else:
-                args.append("--target-os={}".format(self._target_os))
+                args.append(f"--target-os={self._target_os}")
```
Comment (sorny92) [inline recipes/ffmpeg/all/conanfile.py:565]:
```suggestion
                target_os = "Macos" if is_apple_os(self) else str(self.settings.os)
                args.append(f"--target-os={target_os}")
```
So you can get rid of the other `_target_os` or at least the `_get_gnu_triplet`

**Comment by MartinDelille** [comment]
I suppose flac wasn't available since it just has been merged here: https://github.com/conan-io/conan-center-index/pull/15372

I don't know what is the reasonable delay for the package to be available on conan center but let's retrigger the CI just in case.

**Comment by paulharris** [comment]
I see this error in your most recent build failure:
`ERROR: openh264 not found using pkg-config`

I see something similar when build ffmpeg/5.0 with the current CCI recipe, on Windows:
`ERROR: libfdk_aac not found`

libfdk_aac DOES exist, and the .pc file exists in ffmpeg's build folder...
```
$ cat /k/env/conandata/ffmpeg/5.0/_/_/build/a52938edbd2b0d85c72768924adc41517df507a6/fdk-aac.pc
prefix=K:/env/conandata/libfdk_aac/2.0.2/_/_/package/8b3117c31942877c9edf009bf38f3a3b5586c429
libdir=${prefix}/lib
includedir=${prefix}/include

Name: fdk-aac-fdk-aac
Description: Conan package: fdk-aac-fdk-aac
Version: 2.0.2
Libs: -L"${libdir}" -lfdk-aac
Cflags: -I"${includedir}"
```

Pretty sure it was building before!
libfdk_aac's recipe changed recently, perhaps that is the problem.
The Name is a bit weird (fdk-aac-fdk-aac) but I don't normally look at .pc files, perhaps that is normal?

I tried on Proof-of-conan, and it failed with this:
`ERROR: libopenjp2 >= 2.1.0 not found using pkg-config`
https://github.com/eirikb/proof-of-conan/actions/runs/4101301132/jobs/7072951107

**Comment by paulharris** [comment]
Turns out, Proof will work with container = conanio/gcc11-ubuntu16.04 (for the current recipe, not this PR)

File: recipes/ffmpeg/all/conanfile.py [hunk 7]
```diff
@@ -1,18 +1,21 @@
 from conan import ConanFile
 from conan.errors import ConanInvalidConfiguration
-from conan.tools.apple import is_apple_os
+from conan.tools.apple import fix_apple_shared_install_name, is_apple_os, to_apple_arch, XCRun
 from conan.tools.build import cross_building
-from conan.tools.files import get, rename, rmdir
+from conan.tools.files import chdir, copy, get, rename, replace_in_file, rm, rmdir
+from conan.tools.gnu.get_gnu_triplet import _get_gnu_triplet
```
Comment (SpaceIm) [inline recipes/ffmpeg/all/conanfile.py:6]: @MartinDelille no need to trigger c3i again if we know linter will always prevent merge due to this import of a private function.

Comment (MartinDelille) [reply recipes/ffmpeg/all/conanfile.py:6]: Done here: https://github.com/conan-io/conan-center-index/pull/15830

**Comment by prince-chrismc** [comment]
It wont build with merge conflicts :(

**Comment by SpaceIm** [comment]
It's just a waste of CI resource to trigger build of this PR, we know it doesn't pass conan v2 linter, it must be fixed first.

**Pull request closed without merging**